File: backend/src/main/kotlin/com/skewwhiffy/notation/model/Accidental.kt
package com.skewwhiffy.notation.model

data class Accidental(private val offset: Int) {
  companion object {
    val natural: Accidental = Accidental(0)
    /*
    val flat: Accidental = Accidental(-1)
    val sharp: Accidental = Accidental(1)
     */
  }

  val abc: String = when {
    offset == 0 -> ""
    offset < 0 -> "_".repeat(-offset)
    else -> "^".repeat(offset)
  }

  /*

  lazy val displayString: String = offset match {
    case 0 => ""
    case it if it < 0 => "b".repeat(-it)
    case it if it % 2 == 0 => "x".repeat(it / 2)
    case it => "x".repeat((it - 1) / 2) + "#"
  }
*/

  val flat: Accidental
    get() = Accidental(offset - 1)

  val sharp: Accidental
    get() = Accidental(offset + 1)
}

File: backend/src/model/accidental.ts
export interface Accidental {
    readonly offset: number
    readonly abc: string,
    readonly displayString: string
    readonly flat: Accidental
    readonly sharp: Accidental
    equals: (other: Accidental) => boolean
}

export class AccidentalImpl implements Accidental {
    readonly offset: number

    constructor(offset: number) {
        this.offset = offset;
    }

    get abc() {
        if (this.offset < 0) {
            return '_'.repeat(-this.offset);
        }
        return '^'.repeat(this.offset);
    }

    get displayString() {
        if (this.offset < 0) {
            return "b".repeat(-this.offset);
        }
        if (this.offset % 2 == 0) {
            return "x".repeat(this.offset / 2);
        }
        return "x".repeat((this.offset - 1) / 2) + "#";
    }

    get flat(): Accidental {
        return new AccidentalImpl(this.offset - 1);
    }

    get sharp(): Accidental {
        return new AccidentalImpl(this.offset + 1);
    }

    equals(other: Accidental) {
        return other.offset === this.offset
    }
}

type AccidentalTypes = 'natural' | 'flat' | 'sharp'

type AccidentalTypeMap = {
    [key in AccidentalTypes]: Accidental
}

export const accidentalFactory: AccidentalTypeMap = {
    natural: new AccidentalImpl(0),
    flat: new AccidentalImpl(-1),
    sharp: new AccidentalImpl(1),
}
